firebase-firestore/src/main/java/com/google/firebase/firestore/encoding/FirestoreDatatypeSerializers.kt

```diff
@@ -0,0 +1,72 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.firestore.encoding
+
+import com.google.firebase.Timestamp
+import com.google.firebase.firestore.GeoPoint
+import java.util.*
+import kotlinx.serialization.KSerializer
+import kotlinx.serialization.descriptors.SerialDescriptor
+import kotlinx.serialization.descriptors.buildClassSerialDescriptor
+import kotlinx.serialization.encoding.Decoder
+import kotlinx.serialization.encoding.Encoder
+import kotlinx.serialization.modules.SerializersModule
+import kotlinx.serialization.modules.contextual
+
+/**
+ * The generic serializer for data types that need special treatment during (de)serialization
+ * process. The concrete implementation of this generic serializer class is restricted to four
+ * Firestore supported non-primitive types, respectively to be [DocumentReference], [Timestamp], [Date] and [GeoPoint].
+ */
+sealed class FirestoreNativeDataTypeSerializer<T : Any>() : KSerializer<T> {
```

Comment:
Let's not use kotlin for anything we don't strictly have to, and use java instead.

Reply:
I have converted everything else to Java except for this `serializers`, I tried last week and today, if I define the serializers in Java, then I always got `NoSuchElementException` when I try to get the descriptor out in Kotlin.

Reply:
To keep Java library Kotlin file free, we can write Java contextual serializers, and delegate the encoding/decoding process in Kotlin library, but the Java serializer files need to have Kotlin metadata ([Sample Code](https://github.com/firebase/firebase-android-sdk/blob/14ed633b5a9a30c68ef4164d18e5169861b8126d/firebase-firestore/src/main/java/com/google/firebase/firestore/DocumentRefContextualSerializer.java#:~:text=%40Metadata(,%3D%201%2C))).

This looks pretty evil, and as discussed, we are going to keep the Kotlin serializes in the Java library for now.